README.md
[![CI](https://github.com/corgibytes/ruby-debug-ide/actions/workflows/ci.yml/badge.svg)](https://github.com/corgibytes/ruby-debug-ide/actions/workflows/ci.yml)
[![Release Notes](https://github.com/corgibytes/ruby-debug-ide/actions/workflows/release-notes.yml/badge.svg)](https://github.com/corgibytes/ruby-debug-ide/actions/workflows/release-notes.yml)
[![Ruby Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://github.com/testdouble/standard)

# ruby-debug-ide

The ruby-debug-ide gem provides the protocol to establish communication between the debugger engine (such as [debase](https://rubygems.org/gems/debase) or [ruby-debug-base](https://rubygems.org/gems/ruby-debug-base)) and IDEs (for example, [RubyMine](https://www.jetbrains.com/ruby/), [Visual Studio Code](https://code.visualstudio.com/), or [Eclipse](https://www.eclipse.org/ide/)). ruby-debug-ide redirects commands from the IDE to the debugger engine. Then, it returns answers/events received from the debugger engine to the IDE. To learn more about a communication protocol, see the following document: [ruby-debug-ide protocol](protocol-spec.md).

Forked to add support for Docker containers.  See the below issues for more information:

- [#107](https://github.com/ruby-debug/ruby-debug-ide/issues/107) - Docker debugging and sub-debugger with random port
- [#186](https://github.com/ruby-debug/ruby-debug-ide/issues/186) - Error Raised when Debugging a Multi-Process app in Docker on a Mac

## Install

Depending on the used Ruby version you will need the following additional gems installed:

- Ruby 2.x - [debase](https://rubygems.org/gems/debase)

    ```ruby
    group :development, :test do
      gem "ruby-debug-ide", git: "https://github.com/corgibytes/ruby-debug-ide", tag: "v0.7.100.rc1"
      gem "debase"
    end
    ```

- Ruby 1.9.x - [ruby-debug-base19x](https://rubygems.org/gems/ruby-debug-base19x)

    ```ruby
    group :development, :test do
      gem "ruby-debug-ide", git: "https://github.com/corgibytes/ruby-debug-ide", tag: "v0.7.100.rc1"
      gem "ruby-debug-base19x"
    end
    ```

- jRuby or Ruby 1.8.x - [ruby-debug-base](https://rubygems.org/gems/ruby-debug-base)

    ```ruby
    group :development, :test do
      gem "ruby-debug-ide", git: "https://github.com/corgibytes/ruby-debug-ide", tag: "v0.7.100.rc1"
      gem "ruby-debug-base"
    end
    ```

For Windows, make sure that the Ruby [DevKit](https://github.com/oneclick/rubyinstaller/wiki/Development-Kit) is installed.
  
## Using the debugger

There are several different ways to use the ruby-debug-ide gem depending on the IDE you are using and if you are running your application in a container.  Below are some of the setups that we have used succssefully.  If your setup is not mentioned please let us know by opening a [issue](https://github.com/corgibytes/ruby-debug-ide/issues) or [pull request](https://github.com/corgibytes/ruby-debug-ide/pulls).

No matter your setup need to make sure the following ports are open:

- 1234 - Default port that ruby-debug-ide listens on.
- 58430-58450 - Only required if your application spawns subprocesses.  For example, debugging Rails applications that use [Passenger](https://www.phusionpassenger.com/) or [Unicorn](https://yhbt.net/unicorn/) servers.

### Docker Compose, RubyMine, Rails 4 or greater

The `docker-compose.yml` file should look something like:

```yaml
command: bundle exec rails s -p 3000 -b 0.0.0.0

volumes: 
  - .:/sample_rails_app

extra_hosts:
  - "host.docker.internal:host-gateway" # Only required for Linux systems running Docker Desktop.

ports:
  - "3000:3000" # Rails port.
  - "1234:1234" # Port used for by ruby-debug-ide server.
  - "58430-58450:58430-58450" # Sub-process ports for multi-process servers (Unicorn, Passenger, etc).
```

Create a Docker Compose Remote Interpreter as outlined [here](https://www.jetbrains.com/help/ruby/using-docker-compose-as-a-remote-interpreter.html#configure_remote_interpreter).  Open up or create a Rails Run/Debug Configuration as outlined [here](https://www.jetbrains.com/help/ruby/run-rails-applications.html).  Make sure the Ruby SDK used is the Docker Compose one you setup .  

If you are using Docker Desktop make sure you set the Environment Variable to: `IDE_PROCESS_DISPATCHER=host.docker.internal:26162`.  If you don't set the `IDE_PROCESS_DISPATCHER` then ruby-debug-ide will try to respond to the address it receives requests from, which in the Docker world is usually 172.17.0.1.  See [here](https://docs.docker.com/desktop/networking/#i-want-to-connect-from-a-container-to-a-service-on-the-host) for more informaiton on Docker networking.   

On Linux you also need to make sure you set the `extra_hosts` as outlined above.

If you are using just plain old Docker on Linux then you don't need to set the `IDE_PROCESS_DISPATCHER` variable.

The port `26162` is the default port RubyMine uses when spawning the docker container.  For more details see the `docker-compose.override` file created by RubyMine when you click the green debug button.  

```shell
/usr/local/bin/docker-compose -f /home/localadmin/Repos/store/docker-compose.yml -f /home/localadmin/.cache/JetBrains/RubyMine2022.2/tmp/docker-compose.override.2581.yml up --exit-code-from web --abort-on-container-exit web
```

### Docker Compose, RubyMine, Rails 3 or less

Follow the steps in [Docker Compose, RubyMine, Rails 4 or greater](#docker-compose-rubymine-rails-4-or-greater).  The only change is you can't use the default Rails Run/Debug as support for Rails 3.2 was dropped in RubyMine version [2022.1](https://blog.jetbrains.com/ruby/2022/05/rubymine-to-retire-rails-3-and-other-outdated-features/).  Instead you need to create a [Gem Command](https://www.jetbrains.com/help/ruby/run-debug-configuration-gem-command.html) Run/Debug with the following settings:

- Gem name: rails
- Executable name: rails
- Arguments: s -p 3000 -h 0.0.0.0 # Other rails command line arguments as needed.
- Environment variables: IDE_PROCESS_DISPATCHER=host.docker.internal:26162 # If using Docker Desktop

The green RubyMine Run or Debug buttons should behalf as the used too when using the Rails Run/Debug configuration. 

### Command Line and Docker

Run the following command to start the debugging session for a Rails application:

```shell
rdebug-ide --host 0.0.0.0 --port 1234 --dispatcher-port 26162 -- bin/rails s
```

Feel free to add additional options to the rails command at the end such as port, horst, etc.

## ruby-debug-ide Bundled with RubyMine

By default the ruby-debug-ide gem is no longer published to [Rubygems](https://rubygems.org/gems/ruby-debug-ide) as outlined [here](https://github.com/ruby-debug/ruby-debug-ide/issues/201).  Instead it is included with RubyMine at `plugins/ruby/rb/gems`.  Two versions are included with RubyMine: 0.7.3 and 2.3.8.  The 2.3.8 gem, as of this writing, requires Ruby 2.3 or greater the run.

If you don't need to debug older Ruby applications or aren't using Docker you can probably just use the gems bundled with RubyMine.  To do so just list the gem in your Gemfile and RubyMine should do the rest.  For example:

```ruby
group :development, :test do
  gem "ruby-debug-ide"
  gem "debase"
end
```

## Contributing

If you have any questions, notice a bug, or have a suggestion/enhancement please let me know by opening [issue](https://github.com/corgibytes/ruby-debug-ide/issues) or [pull request](https://github.com/corgibytes/ruby-debug-ide/pulls).

### Development Environment Setup

The easiest way to get you development environment setup is to use [Docker](https://www.docker.com/).  [Install Docker](https://docs.docker.com/get-docker/) then run build the container:

```shell
docker-compose build
```

Then run the container:

```bash
docker-compose run --rm app bash
```

Note: The gems are not installed until the container is run and the [docker-entrypoint.sh](docker-entrypoint.sh) is called.

If you don't want to use Docker then you can install Ruby using [RVM](https://rvm.io/), [Rbenv](https://github.com/rbenv/rbenv), or any other method you like.

### Linting

Linting is done by [Standard](https://github.com/testdouble/standard).  To check that the code if formatted correctly run:

```shell
standardrb
```

When working on a file remove it from the [.standard_todo.yml](.standard_todo.yml) file and correct as many linting errors as you can.  Standard is [configured](.standard.yml) to lint the code base for Ruby 1.8 but sometimes it makes recommendations that would break on older Ruby versions.  If you spot a linting recommendation you should open an issue or pull request in [Standard](https://github.com/testdouble/standard). 

### Tests

Once you have your development environment setup make sure the tests all pass:

```bash
rake
```

## Acknowledgements

Thanks to [ruby-debug](https://github.com/ruby-debug) team [JetBrains](https://www.jetbrains.com/) for creating Ruby debugging gems.

CHANGELOG.md
## v0.7.100


As part of this release we had [12 issues](https://github.com/corgibytes/ruby-debug-ide/milestone/1?closed=1) closed.



__DevOps__

- [__#8__](https://github.com/corgibytes/ruby-debug-ide/pull/8) Create Gemfile package in CI
- [__#10__](https://github.com/corgibytes/ruby-debug-ide/pull/10) Generate release notes

__Enhancements__

- [__#1__](https://github.com/corgibytes/ruby-debug-ide/pull/1) Docker container for development
- [__#2__](https://github.com/corgibytes/ruby-debug-ide/pull/2) Create GitHub CI
- [__#3__](https://github.com/corgibytes/ruby-debug-ide/pull/3) Get latest changes from Ruby Debug IDE origin
- [__#4__](https://github.com/corgibytes/ruby-debug-ide/pull/4) Trigger CI on PR
- [__#5__](https://github.com/corgibytes/ruby-debug-ide/pull/5) Trigger CI on push and PR
- [__#6__](https://github.com/corgibytes/ruby-debug-ide/pull/6) Add fixed port range
- [__#7__](https://github.com/corgibytes/ruby-debug-ide/pull/7) Add Standard lint checker
- [__#9__](https://github.com/corgibytes/ruby-debug-ide/pull/9) Fixed some new linting errors with new version of Rubocop
- [__#11__](https://github.com/corgibytes/ruby-debug-ide/pull/11) Rename the gem back to default name
- [__#13__](https://github.com/corgibytes/ruby-debug-ide/pull/13) Allow overriding of dispatcher host and port



lib/ruby-debug-ide/attach/gdb.rb
require 'ruby-debug-ide/attach/native_debugger'

class GDB < NativeDebugger
  def set_flags
    execute 'set scheduler-locking off' # we will deadlock with it
    execute 'set unwindonsignal on'     # in case of some signal we will exit gdb
  end

  def update_threads
    @process_threads = []
    info_threads = (execute 'info threads').split("\n")
    info_threads.each do |thread_info|
      next unless thread_info =~ /[\s*]*\d+\s+Thread.*/
      $stdout.puts "thread_info: #{thread_info}"
      is_main = thread_info[0] == '*'
      thread_num = thread_info.sub(/[\s*]*/, '').sub(/\s.*$/, '').to_i
      thread = ProcessThread.new(thread_num, is_main, thread_info, self)
      if thread.is_main
        @main_thread = thread
      end
      @process_threads << thread
    end
    @process_threads
  end

  def check_already_under_debug
    threads = execute 'info threads'
    threads =~ /ruby-debug-ide/
  end

  def switch_to_thread(thread_num)
    execute "thread #{thread_num}"
  end

  def set_break(str)
    execute "tbreak #{str}"
  end

  def call_start_attach
    super()
    execute "call dlopen(\"#{@path_to_attach}\", 2)"
    execute 'call debase_start_attach()'
    set_break(@tbreak)
  end

  def print_delimiter
    @pipe.puts "print \"#{@delimiter}\""
  end

  def check_delimiter(line)
    line =~ /\$\d+\s=\s"#{@delimiter}"/
  end

  def load_debugger
    execute "call #{@eval_string}"
  end

  def to_s
    GDB.to_s
  end

  class << self
    def to_s
      'gdb'
    end
  end

end


lib/ruby-debug-ide/attach/lldb.rb
require 'ruby-debug-ide/attach/native_debugger'

class LLDB < NativeDebugger
  def set_flags

  end

  def update_threads
    @process_threads = []
    info_threads = (execute 'thread list').split("\n")
    info_threads.each do |thread_info|
      next unless thread_info =~ /[\s*]*thread\s#\d+.*/
      is_main = thread_info[0] == '*'
      thread_num = thread_info.sub(/[\s*]*thread\s#/, '').sub(/:\s.*$/, '').to_i
      thread = ProcessThread.new(thread_num, is_main, thread_info, self)
      if thread.is_main
        @main_thread = thread
      end
      @process_threads << thread
    end
    @process_threads
  end

  def check_already_under_debug
    threads = execute 'thread list'
    threads =~ /ruby-debug-ide/
  end

  def switch_to_thread(thread_num)
    execute "thread select #{thread_num}"
  end

  def set_break(str)
    execute "breakpoint set --shlib #{@path_to_attach} --name #{str}"
  end

  def call_start_attach
    super()
    execute "expr (void *) dlopen(\"#{@path_to_attach}\", 2)"
    execute 'expr (int) debase_start_attach()'
    set_break(@tbreak)
  end

  def print_delimiter
    @pipe.puts "script print \"#{@delimiter}\""
  end

  def check_delimiter(line)
    line =~ /#{@delimiter}$/
  end

  def load_debugger
    execute "expr (void) #{@eval_string}"
  end

  def to_s
    LLDB.to_s
  end

  class << self
    def to_s
      'lldb'
    end
  end

end


lib/ruby-debug-ide/commands/enable.rb
module Debugger
  # Mix-in module to assist in command parsing.
  module EnableDisableFunctions # :nodoc:
    def enable_disable_breakpoints(is_enable, args)
      breakpoints = Debugger.breakpoints.sort_by{|b| b.id }
      largest = breakpoints.inject(0) do |largest_so_far, b|
        b.id if b.id > largest_so_far
      end
      if 0 == largest
        errmsg "No breakpoints have been set.\n"
        return
      end
      args.each do |pos|
        pos = get_int(pos, "#{is_enable} breakpoints", 1, largest)
        return nil unless pos
        breakpoints.each do |b|
          if b.id == pos 
            enabled = ("Enable" == is_enable)
            if enabled
              unless syntax_valid?(b.expr)
                errmsg("Expression \"#{b.expr}\" syntactically incorrect; breakpoint remains disabled.\n")
                break
              end
            end
            b.enabled = enabled
            enabled ? print_breakpoint_enabled(b) : print_breakpoint_disabled(b)
            break
          end
        end
      end
    end

    def enable_disable_display(is_enable, args)
      if 0 == @state.display.size 
        errmsg "No display expressions have been set.\n"
        return
      end
      args.each do |pos|
        pos = get_int(pos, "#{is_enable} display", 1, @state.display.size)
        return nil unless pos
        @state.display[pos-1][0] = ("Enable" == is_enable)
      end
    end

  end

  class EnableCommand < Command # :nodoc:
    Subcommands = 
      [
       ['breakpoints', 2, "Enable specified breakpoints",
"Give breakpoint numbers (separated by spaces) as arguments.
This is used to cancel the effect of the \"disable\" command."
       ],
       ['display', 2, 
        "Enable some expressions to be displayed when program stops",
"Arguments are the code numbers of the expressions to resume displaying.
Do \"info display\" to see current list of code numbers."],
      ].map do |name, min, short_help, long_help| 
      SubcmdStruct.new(name, min, short_help, long_help)
    end unless defined?(Subcommands)

    def regexp
      /^\s* en(?:able)? (?:\s+(.*))?$/ix
    end
    
    def execute
      if not @match[1]
        errmsg "\"enable\" must be followed \"display\", \"breakpoints\"" +
          " or breakpoint numbers.\n"
      else
        args = @match[1].split(/[ \t]+/)
        param = args.shift
        subcmd = find(Subcommands, param)
        if subcmd
          send("enable_#{subcmd.name}", args)
        else
          send(:enable_breakpoints, args.unshift(param))
        end
      end
    end
    
    def enable_breakpoints(args)
      enable_disable_breakpoints("Enable", args)
    end
    
    def enable_display(args)
      enable_disable_display("Enable", args)
    end
    
    class << self
      def help_command
        'enable'
      end

      def help(args)
        if args[1] 
          s = args[1]
          subcmd = Subcommands.find do |try_subcmd| 
            (s.size >= try_subcmd.min) and
              (try_subcmd.name[0..s.size-1] == s)
          end
          if subcmd
            str = subcmd.short_help + '.'
            str += "\n" + subcmd.long_help if subcmd.long_help
            return str
          else
            return "Invalid 'enable' subcommand '#{args[1]}'."
          end
        end
        s = %{
          Enable some things.
          This is used to cancel the effect of the "disable" command.
          -- 
          List of enable subcommands:
          --  
        }
        for subcmd in Subcommands do
          s += "enable #{subcmd.name} -- #{subcmd.short_help}\n"
        end
        return s
      end
    end
  end

  class DisableCommand < Command # :nodoc:
    Subcommands = 
      [
       ['breakpoints', 1, "Disable some breakpoints",
"Arguments are breakpoint numbers with spaces in between.
A disabled breakpoint is not forgotten, but has no effect until reenabled."],
       ['display', 1, "Disable some display expressions when program stops",
"Arguments are the code numbers of the expressions to stop displaying.
Do \"info display\" to see current list of code numbers."],
      ].map do |name, min, short_help, long_help| 
      SubcmdStruct.new(name, min, short_help, long_help)
    end unless defined?(Subcommands)

    def regexp
      /^\s* dis(?:able)? (?:\s+(.*))?$/ix
    end
    
    def execute
      if not @match[1]
        errmsg "\"disable\" must be followed \"display\", \"breakpoints\"" +
          " or breakpoint numbers.\n"
      else
        args = @match[1].split(/[ \t]+/)
        param = args.shift
        subcmd = find(Subcommands, param)
        if subcmd
          send("disable_#{subcmd.name}", args)
        else
          send(:disable_breakpoints, args.unshift(param))
        end
      end
    end
    
    def disable_breakpoints(args)
      enable_disable_breakpoints("Disable", args)
    end
    
    def disable_display(args)
      enable_disable_display("Disable", args)
    end
    
    class << self
      def help_command
        'disable'
      end

      def help(args)
        if args[1] 
          s = args[1]
          subcmd = Subcommands.find do |try_subcmd| 
            (s.size >= try_subcmd.min) and
              (try_subcmd.name[0..s.size-1] == s)
          end
          if subcmd
            str = subcmd.short_help + '.'
            str += "\n" + subcmd.long_help if subcmd.long_help
            return str
          else
            return "Invalid 'disable' subcommand '#{args[1]}'."
          end
        end
        s = %{
          Disable some things.

          A disabled item is not forgotten, but has no effect until reenabled.
          Use the "enable" command to have it take effect again.
          -- 
          List of disable subcommands:
          --  
        }
        for subcmd in Subcommands do
          s += "disable #{subcmd.name} -- #{subcmd.short_help}\n"
        end
        return s
      end
    end
  end

end # module Debugger


lib/ruby-debug-ide.rb
require "pp"
require "stringio"
require "socket"
if RUBY_VERSION < "2.0" || defined?(JRUBY_VERSION)
  require "ruby-debug-base"
  Debugger::FRONT_END = "ruby-debug-base"
else
  require "debase"
  Debugger::FRONT_END = "debase"
end

require "ruby-debug-ide/greeter"
require "ruby-debug-ide/xml_printer"
require "ruby-debug-ide/ide_processor"
require "ruby-debug-ide/event_processor"

module Debugger
  class << self
    ##
    # Finds a free port to use.  The port range is limited to
    # a known range of 58430 to 58450 so Docker/firewall
    # ports can be opened.  Useful when debugging a Rails
    # application that uses multiple process like Unicorn.
    def find_free_port(host)
      # All the possible ports randomly sorted.
      possible_port_numbers = (58430..58450).to_a.shuffle

      # Loop over each port until we find an open port.
      possible_port_numbers.each do |ppn|
        begin
          print_debug("Checking if port is free: #{ppn}\n")

          server = TCPServer.open(host, ppn)
          port = server.addr[1]
          server.close

          # Open port found.
          print_debug("Port #{ppn} is free.\n")
          return port
        rescue
          # Port is in use.  Swallow the error and try
          # another port.
          print_debug("Port #{ppn} is in use.\n")
        end
      end

      # Tried all the ports and couldn't find a free one.
      raise "Unable to find free port."
    end

    # Prints to the stderr using printf(*args) if debug logging flag (-d) is on.
    def print_debug(*args)
      if Debugger.cli_debug
        $stderr.printf("#{Process.pid}: ")
        $stderr.printf(*args)
        $stderr.printf("\n")
        $stderr.flush
      end
    end

    def cleanup_backtrace(backtrace)
      cleared = []
      return cleared unless backtrace
      backtrace.each do |line|
        if line.index(File.expand_path(File.dirname(__FILE__) + "/..")) == 0
          next
        end
        if line.index("-e:1") == 0
          break
        end
        cleared << line
      end
      cleared
    end

    attr_accessor :attached
    attr_accessor :key_value_mode
    attr_accessor :cli_debug, :xml_debug, :evaluation_timeout
    attr_accessor :trace_to_s, :debugger_memory_limit, :inspect_time_limit
    attr_accessor :control_thread
    attr_reader :interface
    # protocol extensions
    attr_accessor :catchpoint_deleted_event, :value_as_nested_element

    #
    # Interrupts the last debugged thread
    #
    def interrupt_last
      skip do
        if (context = last_context)
          return nil unless context.thread.alive?
          context.interrupt
        end
        context
      end
    end

    def start_server(host = nil, port = 1234, notify_dispatcher = false)
      _start_server_common(host, port, nil, notify_dispatcher)
    end

    def start_server_unix(socket_path, notify_dispatcher = false)
      _start_server_common(nil, 0, socket_path, notify_dispatcher)
    end

    def prepare_debugger(options)
      @mutex = Mutex.new
      @proceed = ConditionVariable.new

      if options.socket_path.nil?
        start_server(options.host, options.port, options.notify_dispatcher)
      else
        start_server_unix(options.socket_path, options.notify_dispatcher)
      end

      raise "Control thread did not start (#{@control_thread}}" unless @control_thread && @control_thread.alive?

      # wait for 'start' command
      unless options.skip_wait_for_start
        @mutex.synchronize do
          @proceed.wait(@mutex)
        end
      end
    end

    def debug_program(options)
      prepare_debugger(options)

      abs_prog_script = File.expand_path(Debugger::PROG_SCRIPT)
      bt = debug_load(abs_prog_script, options.stop, options.load_mode)
      if bt && !bt.is_a?(SystemExit)
        $stderr.print "Uncaught exception: #{bt}\n"
        $stderr.print Debugger.cleanup_backtrace(bt.backtrace).map { |l| "\t#{l}" }.join("\n"), "\n"
      end
    end

    def run_prog_script
      return unless @mutex
      @mutex.synchronize do
        @proceed.signal
      end
    end

    def start_control(host, port, notify_dispatcher)
      _start_control_common(host, port, nil, notify_dispatcher)
    end

    def start_control_unix(socket_path, notify_dispatcher)
      _start_control_common(nil, 0, socket_path, notify_dispatcher)
    end

    private

    def _start_server_common(host, port, socket_path, notify_dispatcher)
      return if started?
      start
      _start_control_common(host, port, socket_path, notify_dispatcher)
    end

    def _start_control_common(host, port, socket_path, notify_dispatcher)
      raise "Debugger is not started" unless started?
      return if @control_thread

      @control_thread = DebugThread.new do
        begin
          if socket_path.nil?
            # 127.0.0.1 seemingly works with all systems and with IPv6 as well.
            # "localhost" and nil have problems on some systems.
            # REM
            # host ||= "127.0.0.1"

            server = notify_dispatcher_if_needed(host, port, notify_dispatcher) do |real_port, port_changed|
              s = TCPServer.new(host, real_port)
              print_greeting_msg $stderr, host, real_port, port_changed ? "Subprocess" : "Fast" if defined? IDE_VERSION
              s
            end
          else
            raise "Cannot specify host and socket_file at the same time" if host
            File.delete(socket_path) if File.exist?(socket_path)
            server = UNIXServer.new(socket_path)
            print_greeting_msg $stderr, nil, nil, "Fast", socket_path if defined? IDE_VERSION
          end

          return unless server

          while (session = server.accept)
            if Debugger.cli_debug
              if session.peeraddr == "AF_INET"
                warn "Connected from #{session.peeraddr[2]}"
              else
                warn "Connected from local client"
              end
            end
            dispatcher = ENV["IDE_PROCESS_DISPATCHER"]
            if dispatcher
              ENV["IDE_PROCESS_DISPATCHER"] = "#{session.peeraddr[2]}:#{dispatcher}" unless dispatcher.include?(":")
              ENV["DEBUGGER_HOST"] = host
            end
            begin
              @interface = RemoteInterface.new(session)
              self.handler = EventProcessor.new(interface)
              IdeControlCommandProcessor.new(interface).process_commands
            rescue StandardError, ScriptError => ex
              bt = ex.backtrace
              $stderr.printf "#{Process.pid}: Exception in DebugThread loop: #{ex.message}(#{ex.class})\nBacktrace:\n#{bt ? bt.join("\n  from: ") : "<none>"}\n"
              exit 1
            end
          end
        rescue
          bt = $!.backtrace
          $stderr.printf "Fatal exception in DebugThread loop: #{$!.message}\nBacktrace:\n#{bt ? bt.join("\n  from: ") : "<none>"}\n"
          exit 2
        end
      end
    end

    def notify_dispatcher_if_needed(host, port, need_notify)
      return yield port unless need_notify

      return unless ENV["IDE_PROCESS_DISPATCHER"]
      acceptor_host, acceptor_port = ENV["IDE_PROCESS_DISPATCHER"].split(":")
      # REM
      # acceptor_host, acceptor_port = "127.0.0.1", acceptor_host unless acceptor_port
      connected = false

      3.times do |i|
        unless connected
          begin
            s = TCPSocket.open(acceptor_host, acceptor_port)
            dispatcher_answer = s.gets.chomp

            if dispatcher_answer == "true"
              port = Debugger.find_free_port(host)
            end

            server = yield port, dispatcher_answer == "true"

            s.print(port)
            s.close
            connected = true
            print_debug "Ide process dispatcher notified about sub-debugger which listens on #{port}\n"
            return server
          rescue => bt
            warn "#{Process.pid}: connection failed(#{i + 1})"
            warn "Exception: #{bt}"
            warn bt.backtrace.map { |l| "\t#{l}" }.join("\n")
            sleep 0.3
          end
        end
      end
    end
  end

  class Exception # :nodoc:
    attr_reader :__debug_file, :__debug_line, :__debug_binding, :__debug_context
  end
end
